drivers/meter/ade9153a/ade9153a.c

```diff
@@ -0,0 +1,4418 @@
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "ade9153a.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include "no_os_util.h"
+#include "no_os_spi.h"
+#include "no_os_gpio.h"
+#include "no_os_irq.h"
+#include "no_os_delay.h"
+#include "no_os_units.h"
+#include "no_os_alloc.h"
+#include "no_os_crc16.h"
+#include "no_os_print_log.h"
+#include <stdlib.h>
+#include <errno.h>
+#include <math.h>
+
+NO_OS_DECLARE_CRC16_TABLE(ade9153a_crc16);
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Select comms interface SPI.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9153a_set_interface_spi(struct ade9153a_dev *dev)
+{
+	int ret;
+
+	if (!dev)
+		return -ENOMEM;
+
+	/* reset the ADE and initialize the SPI:
+		SS - Low and SCK - High at reset */
+	if (dev->gpio_ss) {
+		ret = no_os_gpio_direction_output(dev->gpio_ss,
+						  NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+	}
+
+	if (dev->gpio_sck) {
+		ret = no_os_gpio_direction_output(dev->gpio_sck,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+	}
+
+	return ade9153a_hw_reset(dev);
+
+}
+
+/**
+ * @brief Select comms interface Serial.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9153a_set_interface_serial(struct ade9153a_dev *dev)
+{
+	int ret;
+
+	if (!dev)
+		return -ENOMEM;
+
+	/* reset the ADE and initialize the Serial:
+		SS - High and SCK - Low at reset */
+	if (dev->gpio_ss) {
+		ret = no_os_gpio_direction_output(dev->gpio_ss,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+	}
+
+	if (dev->gpio_sck) {
+		ret = no_os_gpio_direction_output(dev->gpio_sck,
+						  NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+	}
+
+	return ade9153a_hw_reset(dev);
+
+}
+
+/**
+ * @brief GPIO interrupt handler for data ready.
+ * @param dev - The device structure.
+ */
+static void ade9153a_irq_handler(void *dev)
+{
+	int ret;
+	/* ADE9153A device descriptor */
+	struct ade9153a_dev *desc = dev;
+	/* data read */
+	uint8_t reg_val;
+
+	/* Disable interrupt while reading data. */
+	ret = no_os_irq_disable(desc->irq_ctrl,
+				desc->gpio_rdy->number);
+	if (ret)
+		return;
+
```

Comment:
isn't there a data read supposed to be happening here ?

Reply:
In the example the data read is not interrupt based. I've left the irq handler in the driver for user applications development. Maybe I should add a comment like this "// user code goes here"